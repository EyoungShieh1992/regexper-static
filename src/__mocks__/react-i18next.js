const reactI18next = jest.requireActual('react-i18next');
const i18n = require('i18n');

module.exports = {
  ...reactI18next,
  Trans: require('__mocks__/component-mock').buildMock(reactI18next.Trans),
  useTranslation: () => ({ i18n, t: i18n.mockT })
};
