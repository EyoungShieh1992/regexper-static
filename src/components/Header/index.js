import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import Modal from 'react-modal';
import { Link } from 'gatsby';
import { Trans } from 'react-i18next';

import GitlabIcon from 'react-feather/dist/icons/gitlab';

import LocaleSwitcher from 'components/LocaleSwitcher';
import InstallPrompt from 'components/InstallPrompt';
import PrivacyPolicy from 'components/PrivacyPolicy';

import style from './style.module.css';

const Header = ({ banner }) => {
  const [ showModal, updateShowModal] = useState(false);
  const handleClose = useCallback(() => {
    updateShowModal(false);
  }, [updateShowModal]);
  const handleOpen = useCallback(event => {
    if (event.shiftKey || event.ctrlKey || event.altKey || event.metaKey) {
      return;
    }

    event.preventDefault();
    updateShowModal(true);
  }, [updateShowModal]);

  return <>
    <Modal
      isOpen={ showModal }
      onRequestClose={ handleClose }>
      <PrivacyPolicy onClose={ handleClose } />
    </Modal>
    <header
      className={ style.header }
      data-banner={ banner || null }>
      <h1>
        <Link to="/">Regexper</Link>
      </h1>

      <ul className={ style.list }>
        <li>
          <a href="https://gitlab.com/javallone/regexper-static"
            rel="external noopener noreferrer"
            target="_blank">
            <GitlabIcon />
            <Trans>Source on GitLab</Trans>
          </a>
        </li>
        <li>
          <Link to="/privacy"
            data-testid="privacy-link"
            onClick={ handleOpen }
          >
            <Trans>Privacy Policy</Trans>
          </Link>
        </li>
        <li>
          <InstallPrompt />
        </li>
        <li data-requires-js>
          <LocaleSwitcher />
        </li>
      </ul>
    </header>
  </>;
};

Header.propTypes = {
  banner: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.string
  ]).isRequired
};

export default Header;
