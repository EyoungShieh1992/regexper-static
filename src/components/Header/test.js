jest.mock('react-modal', () =>
  require('__mocks__/component-mock')('react-modal'));
jest.mock('react-feather/dist/icons/gitlab', () =>
  require('__mocks__/component-mock')('react-feather/dist/icons/gitlab'));
jest.mock('components/LocaleSwitcher', () =>
  require('__mocks__/component-mock')('components/LocaleSwitcher'));
jest.mock('components/InstallPrompt', () =>
  require('__mocks__/component-mock')('components/InstallPrompt'));
jest.mock('components/PrivacyPolicy', () =>
  require('__mocks__/component-mock')('components/PrivacyPolicy'));

import React from 'react';
import { render, fireEvent } from 'react-testing-library';

import Header from 'components/Header';

describe('Header', () => {
  test('rendering', () => {
    const { asFragment } = render(
      <Header banner="testing" />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('rendering with no banner', () => {
    const { asFragment } = render(
      <Header banner={ false } />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('opening the Privacy Policy modal', () => {
    const { asFragment, getByTestId } = render(
      <Header banner={ false } />
    );
    const event = new MouseEvent('click', { bubbles: true });
    jest.spyOn(event, 'preventDefault');

    fireEvent(getByTestId('privacy-link'), event);

    expect(event.preventDefault).toHaveBeenCalled();
    expect(asFragment()).toMatchSnapshot();
  });

  ['shift', 'ctrl', 'alt', 'meta'].forEach(key => {
    test(`opening the Privacy Policy modal while holding ${ key } key`, () => {
      const { asFragment, getByTestId } = render(
        <Header banner={ false } />
      );
      const event = new MouseEvent('click', {
        bubbles: true,
        [key + 'Key']: true
      });
      jest.spyOn(event, 'preventDefault');

      fireEvent(getByTestId('privacy-link'), event);

      expect(event.preventDefault).not.toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
