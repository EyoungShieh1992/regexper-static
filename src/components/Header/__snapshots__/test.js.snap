// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header opening the Privacy Policy modal 1`] = `
<DocumentFragment>
  <span
    data-component="Modal"
    data-props="{
  \\"isOpen\\": true
}"
  >
    <span
      data-component="PrivacyPolicy"
      data-props="{}"
    />
  </span>
  <header
    class="header"
  >
    <h1>
      <a
        href="/"
      >
        Regexper
      </a>
    </h1>
    <ul
      class="list"
    >
      <li>
        <a
          href="https://gitlab.com/javallone/regexper-static"
          rel="external noopener noreferrer"
          target="_blank"
        >
          <span
            data-component="Gitlab"
            data-props="{}"
          />
          <span
            data-component="Trans"
            data-props="{}"
          >
            Source on GitLab
          </span>
        </a>
      </li>
      <li>
        <a
          data-testid="privacy-link"
          href="/privacy"
        >
          <span
            data-component="Trans"
            data-props="{}"
          >
            Privacy Policy
          </span>
        </a>
      </li>
      <li>
        <span
          data-component="InstallPrompt"
          data-props="{}"
        />
      </li>
      <li
        data-requires-js="true"
      >
        <span
          data-component="LocaleSwitcher"
          data-props="{}"
        />
      </li>
    </ul>
  </header>
</DocumentFragment>
`;

exports[`Header opening the Privacy Policy modal while holding alt key 1`] = `
<DocumentFragment>
  <span
    data-component="Modal"
    data-props="{
  \\"isOpen\\": false
}"
  >
    <span
      data-component="PrivacyPolicy"
      data-props="{}"
    />
  </span>
  <header
    class="header"
  >
    <h1>
      <a
        href="/"
      >
        Regexper
      </a>
    </h1>
    <ul
      class="list"
    >
      <li>
        <a
          href="https://gitlab.com/javallone/regexper-static"
          rel="external noopener noreferrer"
          target="_blank"
        >
          <span
            data-component="Gitlab"
            data-props="{}"
          />
          <span
            data-component="Trans"
            data-props="{}"
          >
            Source on GitLab
          </span>
        </a>
      </li>
      <li>
        <a
          data-testid="privacy-link"
          href="/privacy"
        >
          <span
            data-component="Trans"
            data-props="{}"
          >
            Privacy Policy
          </span>
        </a>
      </li>
      <li>
        <span
          data-component="InstallPrompt"
          data-props="{}"
        />
      </li>
      <li
        data-requires-js="true"
      >
        <span
          data-component="LocaleSwitcher"
          data-props="{}"
        />
      </li>
    </ul>
  </header>
</DocumentFragment>
`;

exports[`Header opening the Privacy Policy modal while holding ctrl key 1`] = `
<DocumentFragment>
  <span
    data-component="Modal"
    data-props="{
  \\"isOpen\\": false
}"
  >
    <span
      data-component="PrivacyPolicy"
      data-props="{}"
    />
  </span>
  <header
    class="header"
  >
    <h1>
      <a
        href="/"
      >
        Regexper
      </a>
    </h1>
    <ul
      class="list"
    >
      <li>
        <a
          href="https://gitlab.com/javallone/regexper-static"
          rel="external noopener noreferrer"
          target="_blank"
        >
          <span
            data-component="Gitlab"
            data-props="{}"
          />
          <span
            data-component="Trans"
            data-props="{}"
          >
            Source on GitLab
          </span>
        </a>
      </li>
      <li>
        <a
          data-testid="privacy-link"
          href="/privacy"
        >
          <span
            data-component="Trans"
            data-props="{}"
          >
            Privacy Policy
          </span>
        </a>
      </li>
      <li>
        <span
          data-component="InstallPrompt"
          data-props="{}"
        />
      </li>
      <li
        data-requires-js="true"
      >
        <span
          data-component="LocaleSwitcher"
          data-props="{}"
        />
      </li>
    </ul>
  </header>
</DocumentFragment>
`;

exports[`Header opening the Privacy Policy modal while holding meta key 1`] = `
<DocumentFragment>
  <span
    data-component="Modal"
    data-props="{
  \\"isOpen\\": false
}"
  >
    <span
      data-component="PrivacyPolicy"
      data-props="{}"
    />
  </span>
  <header
    class="header"
  >
    <h1>
      <a
        href="/"
      >
        Regexper
      </a>
    </h1>
    <ul
      class="list"
    >
      <li>
        <a
          href="https://gitlab.com/javallone/regexper-static"
          rel="external noopener noreferrer"
          target="_blank"
        >
          <span
            data-component="Gitlab"
            data-props="{}"
          />
          <span
            data-component="Trans"
            data-props="{}"
          >
            Source on GitLab
          </span>
        </a>
      </li>
      <li>
        <a
          data-testid="privacy-link"
          href="/privacy"
        >
          <span
            data-component="Trans"
            data-props="{}"
          >
            Privacy Policy
          </span>
        </a>
      </li>
      <li>
        <span
          data-component="InstallPrompt"
          data-props="{}"
        />
      </li>
      <li
        data-requires-js="true"
      >
        <span
          data-component="LocaleSwitcher"
          data-props="{}"
        />
      </li>
    </ul>
  </header>
</DocumentFragment>
`;

exports[`Header opening the Privacy Policy modal while holding shift key 1`] = `
<DocumentFragment>
  <span
    data-component="Modal"
    data-props="{
  \\"isOpen\\": false
}"
  >
    <span
      data-component="PrivacyPolicy"
      data-props="{}"
    />
  </span>
  <header
    class="header"
  >
    <h1>
      <a
        href="/"
      >
        Regexper
      </a>
    </h1>
    <ul
      class="list"
    >
      <li>
        <a
          href="https://gitlab.com/javallone/regexper-static"
          rel="external noopener noreferrer"
          target="_blank"
        >
          <span
            data-component="Gitlab"
            data-props="{}"
          />
          <span
            data-component="Trans"
            data-props="{}"
          >
            Source on GitLab
          </span>
        </a>
      </li>
      <li>
        <a
          data-testid="privacy-link"
          href="/privacy"
        >
          <span
            data-component="Trans"
            data-props="{}"
          >
            Privacy Policy
          </span>
        </a>
      </li>
      <li>
        <span
          data-component="InstallPrompt"
          data-props="{}"
        />
      </li>
      <li
        data-requires-js="true"
      >
        <span
          data-component="LocaleSwitcher"
          data-props="{}"
        />
      </li>
    </ul>
  </header>
</DocumentFragment>
`;

exports[`Header rendering 1`] = `
<DocumentFragment>
  <span
    data-component="Modal"
    data-props="{
  \\"isOpen\\": false
}"
  >
    <span
      data-component="PrivacyPolicy"
      data-props="{}"
    />
  </span>
  <header
    class="header"
    data-banner="testing"
  >
    <h1>
      <a
        href="/"
      >
        Regexper
      </a>
    </h1>
    <ul
      class="list"
    >
      <li>
        <a
          href="https://gitlab.com/javallone/regexper-static"
          rel="external noopener noreferrer"
          target="_blank"
        >
          <span
            data-component="Gitlab"
            data-props="{}"
          />
          <span
            data-component="Trans"
            data-props="{}"
          >
            Source on GitLab
          </span>
        </a>
      </li>
      <li>
        <a
          data-testid="privacy-link"
          href="/privacy"
        >
          <span
            data-component="Trans"
            data-props="{}"
          >
            Privacy Policy
          </span>
        </a>
      </li>
      <li>
        <span
          data-component="InstallPrompt"
          data-props="{}"
        />
      </li>
      <li
        data-requires-js="true"
      >
        <span
          data-component="LocaleSwitcher"
          data-props="{}"
        />
      </li>
    </ul>
  </header>
</DocumentFragment>
`;

exports[`Header rendering with no banner 1`] = `
<DocumentFragment>
  <span
    data-component="Modal"
    data-props="{
  \\"isOpen\\": false
}"
  >
    <span
      data-component="PrivacyPolicy"
      data-props="{}"
    />
  </span>
  <header
    class="header"
  >
    <h1>
      <a
        href="/"
      >
        Regexper
      </a>
    </h1>
    <ul
      class="list"
    >
      <li>
        <a
          href="https://gitlab.com/javallone/regexper-static"
          rel="external noopener noreferrer"
          target="_blank"
        >
          <span
            data-component="Gitlab"
            data-props="{}"
          />
          <span
            data-component="Trans"
            data-props="{}"
          >
            Source on GitLab
          </span>
        </a>
      </li>
      <li>
        <a
          data-testid="privacy-link"
          href="/privacy"
        >
          <span
            data-component="Trans"
            data-props="{}"
          >
            Privacy Policy
          </span>
        </a>
      </li>
      <li>
        <span
          data-component="InstallPrompt"
          data-props="{}"
        />
      </li>
      <li
        data-requires-js="true"
      >
        <span
          data-component="LocaleSwitcher"
          data-props="{}"
        />
      </li>
    </ul>
  </header>
</DocumentFragment>
`;
